torch/csrc/jit/passes/quantization.cpp

```diff
@@ -88,7 +158,37 @@ void InsertObserverNodes(std::shared_ptr<Graph>& graph, Node* observer_node) {
 }
 
 void InsertFakeQuantNodes(std::shared_ptr<Graph>& graph) {
-  throw std::runtime_error("Pass not implemented yet!");
+  std::stack<Block*> blocks_to_visit;
+  blocks_to_visit.push(graph->block());
+  while (!blocks_to_visit.empty()) {
+    Block* b = blocks_to_visit.top();
+    blocks_to_visit.pop();
+
+    for (Node* n : b->nodes()) {
+      for (Block* subblock : n->blocks()) {
+        blocks_to_visit.push(subblock);
+      }
+
+      if (!checkIfNodeValuesQuantizable(n)) {
+        continue;
+      }
+      // Basic notion is for a quantizable node to inspect inputs and outputs
+      // We quantize all the outputs
+      // We selectively quantize the inputs depending on parent node
+      for (auto& v : n->outputs()) {
+        if (isTypeTensor(v)) {
+          addQuantDeQuantNodes(v);
+        }
+      }
+      // This is to quantize external inputs. Nodes for external input
+      // do not belong to any block
+      for (auto& v : n->inputs()) {
```

Comment:
Why not loop over the graph/block inputs explicitly instead of finding them as inputs?
Isn't this going to run multiple times on Graph/ block inputs?

Reply:
Yeah i considered that approach. For that i need to iterate over v->uses and modify if i identify use as a quantizable node. Documentation said any change to v will invalidate the use iterator. Else  i need to accumulate quantizable uses for v first, and then modify graph. Please suggest